smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/waiters/JMESPathVisitor.kt

```diff
@@ -0,0 +1,232 @@
+            writer.write("let current = Optional.some(current)")
+            val comparisonName = childBlock(expression.comparison!!)
+            writer.write("return \$L", comparisonName)
+        }
+        val right = expression.right!!
+        return mappingBlock(right, filteredName)
+    }
+
+    // Returns the inner expression unchanged when the inner expression is an array of non-array elements.
+    // Returns the inner expression flattened when the inner expression is an array of arrays.
+    override fun visitFlatten(expression: FlattenExpression): String {
+        val innerName = expression.expression!!.accept(this)
+        return addTempVar("${innerName}OrEmpty", "Optional.some((\$L ?? []).flattenIfPossible { $$0 })", innerName)
+    }
+
+    // Implement contains() and length() free functions which are the only 2 JMESPath methods we support.
+    // contains() returns true if its 1st param is a collection that contains an element equal
+    // to the 2nd param, false otherwise.
+    // length() returns the number of elements of an array, the number of key/value pairs for a map,
+    // or the number of characters for a string.
+    override fun visitFunction(expression: FunctionExpression): String {
+        when (expression.name) {
+            "contains" -> {
+                if (expression.arguments.size != 2) { throw Exception("Unexpected number of arguments to $expression") }
+
+                val subject = expression.arguments[0]
+                val subjectName = subject.accept(this)
+                val search = expression.arguments[1]
+                val searchName = search.accept(this)
+                return addTempVar("contains", "\$L.flatMap { \$L?.contains($$0) } ?? false", searchName, subjectName)
+            }
+            "length" -> {
+                if (expression.arguments.size != 1) { throw Exception("Unexpected number of arguments to $expression") }
+
+                val subject = expression.arguments[0]
+                val subjectName = subject.accept(this)
+                return addTempVar("count", "Optional.some(Double(\$L?.count ?? 0))", subjectName)
+            }
+            else -> throw Exception("Unknown function type in $expression")
+        }
+    }
+
+    override fun visitIndex(expression: IndexExpression): String {
+        throw Exception("IndexExpression is unsupported")
+    }
+
+    // Renders a literal of any supported type, wrapped in an optional.
+    override fun visitLiteral(expression: LiteralExpression): String {
+        when (expression.type) {
+            RuntimeType.STRING -> return addTempVar("string", "Optional.some(\"\$L\")", expression.expectStringValue())
+            RuntimeType.NUMBER -> return addTempVar("number", "Optional.some(Double(\$L))", expression.expectNumberValue())
+            RuntimeType.BOOLEAN -> return addTempVar("bool", "Optional.some(\$L)", expression.expectBooleanValue())
```

Comment:
It looks like we don't need to wrap these in optionals since the result is always passed into a `JMESValue`. It also feels a bit off to be wrapping literals in optionals when we know they aren't optional.